src/sentry/utils/pytest/fixtures/bad_model_signal_handlers.py

```diff
@@ -0,0 +1,96 @@
+import pytest
+
+
+def _format_msg(bad_signal_handlers):
+    msg = """\
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+We have detected that you are spawning a celery task inside of a model signal
+handler such as post_save or post_delete without using transaction.on_commit:
+"""
+
+    for signal_sender, task_self in bad_signal_handlers:
+        msg += f"  - A change to model {signal_sender} spawning a task {task_self.__name__}\n"
+
+    msg += """
+We saw that this kind of signal handler is often subtly broken in how it
+interacts with database transactions. Typically the fix is to change the signal
+handler to use django.db.transaction.on_commit:
+
+    def my_signal_handler(...):
+        # Before
+        handle_model_changes.apply_async(...)
+        # After
+        transaction.on_commit(lambda: handle_model_changes.apply_async(...))
+
+For an example of a real-world fix, see
+https://github.com/getsentry/sentry/pull/35523, and for the PR that introduced
+this fixture see https://github.com/getsentry/sentry/pull/35671
+
+If you think this doesn't apply to this test, use the
+`bad_model_signal_handlers` fixture like so at the end of the test:
+
+    class MyTestCase(TestCase):
+        def test_very_special(self):
+            self.bad_model_signal_handlers.clear()
+
+Or like this in pytest-based tests:
+
+    def test_very_special(bad_model_signal_handlers):
+        # Do something odd with models here
+
+        bad_model_signal_handlers.clear()
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"""
+
+    return msg
+
+
+@pytest.fixture(autouse=True)
+def bad_model_signal_handlers(monkeypatch):
```

Comment:
Just had a thought - does it make sense to also check whether code is inside a `transaction.atomic` context manager? The problem we're really trying to avoid here is firing a task from inside an uncommitted transaction, rather than just firing from signal handlers.

Reply:
@wedamija my thinking was that any signal handler for models that triggers a celery task needs to be resilient against hypothetical transactions, regardless of whether transactions _actually happen_ right now.

But the more practical problem for doing this check is that all the tests actually already run in a transaction.atomic block, because that's how test isolation works.

Reply:
I wasn't sure if the tests are actually running in a `transaction.atomic` block. It does make things trickier, but I think we can handle it. We'd just need to moneypatch differently in normal tests vs transaction tests.

We can increment a counter whenever we enter a `transaction.atomic` block, and decrement when we exit. Then in normal tests, if the counter is above 1 then we can assume it's a real transaction and warn. Transaction tests will be the same, but check whether the counter is above 0.

Reply:
Yeah that makes sense. Are you sure we want to only reject signal handlers under that condition though? I think it would make sense to establish a general rule that model signal handlers need to handle transactions correctly.

Reply:
I think we can do both. What you're doing is good since it will catch any signal handlers that are spawning tasks, regardless of whether they're in a transaction. 

But I think it would also be helpful to catch other uses of transactions that are spawning tasks. Something like:
 ```
 with transaction.atomic():
    my_instance = Model.objects.get(...)
    # do stuff
    my_instance.save()
    task_that_relies_on_instance.delay(id=my_instance.id)
    # more stuff
```

Reply:
_i see_. i thought you were proposing restricting the existing check to only fire if the model is being saved within a transaction blog. I'll look into adding this other check tomorrow